crossbar/worker/proxy.py

```diff
@@ -89,352 +363,140 @@ def onConnect(self):
-        endpoint = _create_transport_endpoint(reactor, backend_config["endpoint"])
-        self._transport_d = endpoint.connect(factory)
-
-        def good(proto):
-            self._transport_d = None
-            # we give the backend a way to get back to us .. perhaps
-            # there's another / better way?
-            proto._proxy_other_side = self
-            self._backend_transport = proto
-
-            # So, we need to wait for the connection to "be open"
-            # before we can forward the Hello .. but there's no
-            # listener interface on RawSocket :/ so we have to wrap
-            # things ... at least we know "it's all Twisted" here (but
-            # also: perhaps there SHOULD be a more-general interface
-            # for this)
-
-            if not proto.isOpen():
-                # XXX rawsocket-specific
-                if hasattr(proto, "_on_handshake_complete"):
-                    orig = proto._on_handshake_complete
-
-                    def _connected(*args, **kw):
-                        x = orig(*args, **kw)
-                        self._backend_transport.send(hello_msg)
-                        return x
-                    proto._on_handshake_complete = _connected
-
-                    def _closed(*arg, **kw):
-                        self.close()
-                    proto.is_closed.addCallback(_closed)
-
-                # XXX websocket-specific
-                else:
-
-                    def _connected(arg):
-                        self._backend_transport.send(hello_msg)
-                        return arg
-                    proto.is_open.addCallback(_connected)
-
-                    def _closed(*arg, **kw):
-                        self.close()
-                    proto.is_closed.addCallback(_closed)
-            else:
-                self._backend_transport.send(hello_msg)
-
-        def bad(f):
-            self._transport_d = None
-            print("fail: {}".format(f))
-            self._teardown()
-        self._transport_d.addCallbacks(good, bad)
-        return self._transport_d
-
-
-class ProxyWebSocketService(RouterWebService):
-    """
-    For every 'type=websocket-proxy' node configured in a proxy there is one
-    of these; it will start FrontendProxySession instances upon every client
-    connection.
-    """
-    _backend_configs = None
+    def _error(f):
+        if not connected_d.called:
+            connected_d.errback(f)
+    transport_d.addErrback(_error)
+    transport_d.addCallback(_connected)
 
-    @staticmethod
-    def create(transport, path, config, controller):
-        websocket_factory = WebSocketServerFactory()
-        websocket_factory.protocol = FrontendProxyProtocol
+    return connected_d
 
-        resource = WebSocketResource(websocket_factory)
 
-        service = ProxyWebSocketService(transport, path, config, resource)
-        websocket_factory._service = service
+# implements IRealmContainer
+class ProxyController(RouterController):
```

Comment:
fantastic! can_map_backend and map_backend and auxiliary functions, this makes a lot of sense and nicely factors the code with little binding in between. I like that a lot .. it just seams natural from now looking over the implementation.